package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,943 @@
+        bbox_vol = box.c_pbcbox.box[XX][XX] * box.c_pbcbox.box[YY][YY] * box.c_pbcbox.box[YY][YY]
+        size = bbox_vol/cutoff**3
+        nbeadspercell = ncoords/size
+        while bbox_vol/self.cutoff**3 > max_size:
+            self.cutoff *= 1.2
+
+        for i in range(DIM):
+            self.ncells[i] = <ns_int> (box.c_pbcbox.box[i][i] / self.cutoff)
+            if self.ncells[i] == 0:
+                self.ncells[i] = 1
+            self.cellsize[i] = box.c_pbcbox.box[i][i] / self.ncells[i]
+        self.size = self.ncells[XX] * self.ncells[YY] * self.ncells[ZZ]
+
+        if self.debug:
+            print("NSGrid: Requested cutoff: {:.3f} (Ncells={}, Avg # of beads per cell={}), Optimized cutoff= {:.3f} (Ncells={}, Avg # of beads per cell={})".format(
+                cutoff, size, nbeadspercell,
+                self.cutoff, self.size, <ns_int> (ncoords / self.size)
+            ))
+            print("NSGrid: Size={}x{}x{}={}".format(self.ncells[XX], self.ncells[YY], self.ncells[ZZ], self.size))
+
+        self.cell_offsets[XX] = 0
+        self.cell_offsets[YY] = self.ncells[XX]
+        self.cell_offsets[ZZ] = self.ncells[XX] * self.ncells[YY]
+
+        # Allocate memory
+        self.nbeads = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size)
+        if not self.nbeads:
+            raise MemoryError("Could not allocate memory from NSGrid.nbeads ({} bits requested)".format(sizeof(ns_int) * self.size))
+        self.beadids = NULL
+        self.cellids = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.ncoords)
+        if not self.cellids:
+            raise MemoryError("Could not allocate memory from NSGrid.cellids ({} bits requested)".format(sizeof(ns_int) * self.ncoords))
+        self.nbeads_per_cell = 0
+
+        for i in range(self.size):
+            self.nbeads[i] = 0
+
+    def __dealloc__(self):
+        PyMem_Free(self.nbeads)
+        PyMem_Free(self.beadids)
+        PyMem_Free(self.cellids)
+
+    cdef ns_int coord2cellid(self, rvec coord) nogil:
+        return <ns_int> (coord[ZZ] / self.cellsize[ZZ]) * (self.ncells[XX] * self.ncells[YY]) +\
+               <ns_int> (coord[YY] / self.cellsize[YY]) * self.ncells[XX] + \
+               <ns_int> (coord[XX] / self.cellsize[XX])
+
+    cdef bint cellid2cellxyz(self, ns_int cellid, ivec cellxyz) nogil:
+        if cellid < 0:
+            return False
+        if cellid >= self.size:
+            return False
+
+        cellxyz[ZZ] = <ns_int> (cellid / self.cell_offsets[ZZ])
+        cellid -= cellxyz[ZZ] * self.cell_offsets[ZZ]
+
+        cellxyz[YY] = <ns_int> (cellid / self.cell_offsets[YY])
+        cellxyz[XX] = cellid - cellxyz[YY] * self.cell_offsets[YY]
+
+        return True
+
+    cdef fill_grid(self, real[:, ::1] coords):
+        cdef ns_int i, cellindex = -1
+        cdef ns_int ncoords = coords.shape[0]
+        cdef ns_int *beadcounts = NULL
+
+        # Allocate memory
+        beadcounts = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size)

```

Comment:
I would replace this with `cdef ns_int_t[:] beadcounts = np.empty(self.size, dtype=ns_int)`. Then python does proper reference counting and you avoid the memory error mentioned earlier.

Reply:
My example code assumes that you have something like
```
ns_int_t = np.int64_t
ns_int = np.int64
```
at the beginning of the file to distinguish between the int type and the numpy variable. It's common convention also used in the cython docs and examples.